src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java

```diff
@@ -0,0 +1,438 @@
+            throw new RequestException(ErrorType.NO_VALID_HOST_FOUND, ResponseConstants.NO_VALID_HOST_FOUND);
+        } else if (response.getStatusLine().getStatusCode() > 204) {
+            throw new RequestException(ErrorType.BAD_REQUEST,
+                    "Status code: " + response.getStatusLine().toString() + " | Message:" + message);
+        }
+    }
+
+    protected Flavor updateAndFindSmallestFlavor(ComputeOrder computeOrder) {
+        updateFlavors(computeOrder.getLocalToken());
+
+        return findSmallestFlavor(computeOrder);
+    }
+
+    private Flavor findSmallestFlavor(ComputeOrder computeOrder) {
+        List<Flavor> listFlavor = new ArrayList<>();
+        for (Flavor flavor : flavors) {
+            if (matches(flavor, computeOrder)) {
+                listFlavor.add(flavor);
+            }
+        }
+
+        if (listFlavor.isEmpty()) {
+            return null;
+        }
+
+        Collections.sort(listFlavor);
+
+        return listFlavor.get(0);
+    }
+
+    private boolean matches(Flavor flavor, ComputeOrder computeOrder) {
+        if (flavor.getDisk() < computeOrder.getDisk()
+                || flavor.getCpu() < computeOrder.getvCPU()
+                || flavor.getMem() < computeOrder.getMemory()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    protected synchronized void updateFlavors(Token localToken) {
+        try {
+            String tenantId = getTenantId(localToken);
+            if (tenantId == null) {
+                return;
+            }
+
+            String endpoint = getComputeEndpoint(tenantId, SUFFIX_ENDPOINT_FLAVORS);
+            String jsonResponseFlavors = doGetRequest(endpoint, localToken);
+
+            Map<String, String> nameToFlavorId = new HashMap<>();
+
+            JSONArray jsonArrayFlavors = new JSONObject(jsonResponseFlavors).getJSONArray("flavors");
+
+            for (int i = 0; i < jsonArrayFlavors.length(); i++) {
+                JSONObject itemFlavor = jsonArrayFlavors.getJSONObject(i);
+                nameToFlavorId.put(itemFlavor.getString(NAME_JSON_FIELD), itemFlavor.getString(ID_JSON_FIELD));
+            }
+
+            List<Flavor> newFlavors = detailFlavors(endpoint, localToken, nameToFlavorId);
+            if (newFlavors != null) {
+                this.flavors.addAll(newFlavors);
+            }
+
+            removeInvalidFlavors(nameToFlavorId);
+
+        } catch (Exception e) {
+            LOGGER.warn("Error while updating flavors.", e);
+        }
+    }
+
+    private List<Flavor> detailFlavors(String endpoint, Token localToken,
+                                       Map<String, String> nameToIdFlavor) throws JSONException, RequestException {
+        List<Flavor> newFlavors = new ArrayList<>();
+        List<Flavor> flavorsCopy = new ArrayList<>(flavors);
+
+        for (String flavorName : nameToIdFlavor.keySet()) {
+            boolean containsFlavor = false;
+            for (Flavor flavor : flavorsCopy) {
```

Comment:
These **for** seems unnecessary to me, a right implementation of the Flavor equals and hashCode methods can substitute these **for** by a **contains** TreeSet method.